src/partial-assembly.md

```diff
@@ -0,0 +1,174 @@
+# Performance and Partial Assembly
+
+This document provides a brief overview of the tensor-based high-performance
+and partial assembly features in MFEM. In the traditional finite element
+setting, the operator is assembled in the form of a matrix. The action of the
+operator is computed by multiplying with this matrix. At high orders this
+requires both a large amount of memory to store the matrix, as well as many
+floating point operations to compute and apply it. _Partial assembly_ is a
+technique that allows for efficiently applying the action of finite element
+operators without forming the corresponding matrix. This is particularly
+important when running on GPUs.
+
+Partial assembly is enabled at the level of the `BilinearForm` by setting
+the assembly level:
+```c++
+a->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+```
+Once partial assembly is enabled, subsequent calls to member functions such as
+`FormLinearSystem` will result in an `Operator` that represents the action of
+the bilinear form `a`, without assembling a matrix. This functionality is
+illustrated in several [MFEM examples](examples.md), including example 1, 3, 6,
+9, 24, and 26.
+
+Note that partial assembly is currently implemented for tensor-product elements
+(i.e. quadrilaterals and hexahedra). Partial assembly for simplex elements
+(triangles and tetrahedra) is planned.
+
+## Preconditioning with Partial Assembly
+
+When using partial assembly, the system matrix is no longer available for
+constructing preconditioners. This means that some of the standard
+preconditioners in MFEM such as `HypreBoomerAMG` and `GSSmoother` cannot be
+used.
+
+MFEM allows for the efficient construction of diagonal (Jacobi) smoothers for
+partially assembled operators on quad and hex meshes using the class
+`OperatorJacobiSmoother`. This class efficiently assembles the diagonal of the
+corresponding matrix, exploiting the tensor-product structure for efficient
+evaluation.
+
+MFEM also allows for [Chebyshev
+smoothing](http://netlib.org/linalg/html_templates/node76.html) with partial
+assembly using the class `OperatorChebyshevSmoother`. This smoother uses
+estimates of the eigenvalues of the operator computed using the [power
+method](https://en.wikipedia.org/wiki/Power_iteration), and is built upon the
+functionality of `OperatorJacobiSmoother`.
+
+These smoothers can be used within the context of h- and p-multigrid methods.
```

Comment:
I feel like this paragraph undersells the multigrid methods, which are probably more important than just the smoothers for most users. Maybe something like "Very efficient" (optimal?) "h-multigrid and p-multigrid preconditioners can be constructed for PA operators by leveraging a hierarchy of discretizations and the smoothers described above..."

Reply:
👍